File: crates/dw-ado/src/changelog.rs
use crate::{
    AdoError, AdoToken, AzureDevOpsOptions, WorkItemSnapshot, get_related_work_item_ids,
    get_work_item_snapshot_authenticated, get_work_item_snapshots_authenticated,
    try_get_pull_request_work_item_ids,
};
use dw_core::{PullRequestId, WorkItemId};
use serde::{Deserialize, Serialize};
use std::collections::{BTreeMap, BTreeSet};

pub const RELATION_HIERARCHY_REVERSE: &str = "System.LinkTypes.Hierarchy-Reverse";
pub const RELATION_HIERARCHY_FORWARD: &str = "System.LinkTypes.Hierarchy-Forward";

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct WorkItemGroup {
    pub parent: WorkItemSnapshot,
    pub items: Vec<WorkItemSnapshot>,
}

pub fn extract_work_item_ids_from_commit_messages(commit_log: &str) -> Vec<WorkItemId> {
    let mut ids = Vec::new();
    for (index, _) in commit_log.match_indices('#') {
        let id = commit_log[index + 1..]
            .chars()
            .take_while(|ch| ch.is_ascii_digit())
            .collect::<String>();
        let id = WorkItemId::from(id);
        if !id.as_str().is_empty() && !ids.iter().any(|existing| existing == &id) {
            ids.push(id);
        }
    }
    ids
}

pub fn group_work_items_by_parent(
    options: &AzureDevOpsOptions,
    items: &[WorkItemSnapshot],
    token: &AdoToken,
) -> Result<Vec<WorkItemGroup>, AdoError> {
    let mut groups: BTreeMap<WorkItemId, Vec<WorkItemSnapshot>> = BTreeMap::new();
    let mut parents: BTreeMap<WorkItemId, WorkItemSnapshot> = BTreeMap::new();

    for item in items {
        let parent_id = get_related_work_item_ids(
            options,
            item.id.as_str(),
            RELATION_HIERARCHY_REVERSE,
            token,
        )?
        .into_iter()
        .next()
        .unwrap_or_else(|| item.id.clone());
        if parent_id == item.id {
            parents.insert(parent_id.clone(), item.clone());
        } else if !parents.contains_key(&parent_id) {
            parents.insert(
                parent_id.clone(),
                get_work_item_snapshot_authenticated(options, parent_id.as_str(), token)?,
            );
        }

        let children = groups.entry(parent_id.clone()).or_default();
        if parent_id != item.id {
            children.push(item.clone());
        }
    }

    Ok(groups
        .into_iter()
        .filter_map(|(parent_id, mut items)| {
            items.sort_by(|left, right| left.id.cmp(&right.id));
            Some(WorkItemGroup {
                parent: parents.remove(&parent_id)?,
                items,
            })
        })
        .collect())
}

pub fn get_work_item_ids_from_pull_requests(
    options: &AzureDevOpsOptions,
    repositories: &[String],
    pull_request_ids: &[PullRequestId],
    token: &AdoToken,
) -> Result<Vec<WorkItemId>, AdoError> {
    if repositories.is_empty() {
        return Err(AdoError::InvalidInput(
            "PR mode requires an explicit repository, or a project with configured AzureDevOpsRepository entries.".into(),
        ));
    }

    let mut ids = Vec::new();
    for pull_request_id in pull_request_ids {
        let numeric_pull_request_id = pull_request_id.as_str().parse::<i64>().map_err(|_| {
            AdoError::InvalidInput(format!("Invalid pull request ID: {pull_request_id}"))
        })?;
        let mut matches = Vec::new();
        for repository in repositories {
            if let Some(work_item_ids) = try_get_pull_request_work_item_ids(
                options,
                repository,
                numeric_pull_request_id,
                token,
            )? {
                matches.push((repository.clone(), work_item_ids));
            }
        }

        match matches.len() {
            0 => {
                return Err(AdoError::Request(format!(
                    "Pull request #{pull_request_id} was not found in tested Azure DevOps repos: {}",
                    repositories.join(", ")
                )));
            }
            1 => ids.extend(matches.remove(0).1),
            _ => {
                return Err(AdoError::InvalidInput(format!(
                    "Pull request #{pull_request_id} was found in multiple repos ({}). Specify the repository.",
                    matches
                        .into_iter()
                        .map(|item| item.0)
                        .collect::<Vec<_>>()
                        .join(", ")
                )));
            }
        }
    }

    let mut seen = BTreeSet::new();
    Ok(ids
        .into_iter()
        .filter(|id| seen.insert(id.clone()))
        .collect())
}

pub fn load_changelog_items(
    options: &AzureDevOpsOptions,
    work_item_ids: &[WorkItemId],
    token: &AdoToken,
) -> Result<Vec<WorkItemSnapshot>, AdoError> {
    get_work_item_snapshots_authenticated(options, work_item_ids, token)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extract_work_item_ids_from_commit_messages_reads_ids_in_order_and_dedupes() {
        let commit_log = "fix(#104 #105): corriger le calcul\u{1e}refactor(#104 #107): simplifier";

        let ids = extract_work_item_ids_from_commit_messages(commit_log);

        assert_eq!(
            ids,
            vec![
                WorkItemId::from("104"),
                WorkItemId::from("105"),
                WorkItemId::from("107")
            ]
        );
        assert_eq!(
            ids.iter()
                .map(ToString::to_string)
                .collect::<Vec<_>>()
                .join(" "),
            "104 105 107"
        );
    }
}

File: internal/work/ado/changelog.go
package ado

import (
	"context"
	"sort"
	"strconv"
	"strings"

	"github.com/sachahjkl/dw/internal/work"
)

const maximumWorkItemIDDigits = 7

func ExtractWorkItemIDsFromCommitMessages(commitLog string) []string {
	result := make([]string, 0)
	seen := make(map[string]struct{})
	for offset := 0; offset < len(commitLog); {
		index := strings.IndexByte(commitLog[offset:], '#')
		if index < 0 {
			break
		}
		hash := offset + index
		start := hash + 1
		end := start
		for end < len(commitLog) && isASCIIDigit(commitLog[end]) {
			end++
		}
		offset = max(end, start)
		if end == start || end-start > maximumWorkItemIDDigits || !workItemReferenceAt(commitLog, hash, end) {
			continue
		}
		id := strings.TrimLeft(commitLog[start:end], "0")
		if id == "" {
			continue
		}
		if _, exists := seen[id]; !exists {
			seen[id] = struct{}{}
			result = append(result, id)
		}
	}
	return result
}

func workItemReferenceAt(text string, hash, end int) bool {
	if end < len(text) && (isWordByte(text[end]) || text[end] == ';') {
		return false
	}
	if hash == 0 {
		return true
	}
	previous := text[hash-1]
	switch {
	case hash >= 2 && strings.EqualFold(text[hash-2:hash], "AB"):
		return hash == 2 || !isWordByte(text[hash-3])
	case isWordByte(previous), previous == '&', previous == '=', previous == '"', previous == '\'', previous == '/':
		return false
	case previous == '(':
		githubPullRequest := (hash == 1 || isSpaceByte(text[hash-2])) && end < len(text) && text[end] == ')'
		return !githubPullRequest
	}
	return true
}

func isASCIIDigit(value byte) bool { return value >= '0' && value <= '9' }

func isWordByte(value byte) bool {
	return isASCIIDigit(value) || value == '_' || (value >= 'a' && value <= 'z') || (value >= 'A' && value <= 'Z') || value >= 0x80
}

func isSpaceByte(value byte) bool {
	return value == ' ' || value == '\t' || value == '\n' || value == '\r'
}

func (*Provider) ExtractCommitReferences(commitLog string) []work.ItemID {
	raw := ExtractWorkItemIDsFromCommitMessages(commitLog)
	result := make([]work.ItemID, len(raw))
	for index := range raw {
		result[index] = work.ItemID(raw[index])
	}
	return result
}

func (p *Provider) GroupWorkItemsByParent(ctx context.Context, options Options, items []WorkItemSnapshot, token Token) ([]WorkItemGroup, error) {
	ids := make([]string, len(items))
	for index, item := range items {
		ids[index] = item.ID
	}
	objects, err := p.workItemObjects(ctx, options, ids, true, token)
	if err != nil {
		return nil, err
	}
	parentOf := make(map[string]string, len(objects))
	for _, value := range objects {
		if id, parent := elementText(value["id"]), parentIDFromObject(value); id != nil && parent != nil {
			parentOf[*id] = *parent
		}
	}
	parentIDs := make([]string, 0)
	for _, item := range items {
		if parentID, found := parentOf[item.ID]; found && !containsText(parentIDs, parentID) {
			parentIDs = append(parentIDs, parentID)
		}
	}
	loaded, err := p.GetWorkItemsBatch(ctx, options, parentIDs, token)
	if err != nil {
		return nil, err
	}
	parents := make(map[string]WorkItemSnapshot, len(loaded))
	for _, parent := range loaded {
		parents[parent.ID] = parent
	}
	groups := make(map[string][]WorkItemSnapshot)
	orphans := make([]WorkItemSnapshot, 0)
	for _, item := range items {
		parentID, found := parentOf[item.ID]
		if !found {
			if _, exists := groups[item.ID]; !exists {
				groups[item.ID] = make([]WorkItemSnapshot, 0)
			}
			parents[item.ID] = item
			continue
		}
		if _, loaded := parents[parentID]; !loaded {
			orphans = append(orphans, item)
			continue
		}
		groups[parentID] = append(groups[parentID], item)
	}
	keys := make([]string, 0, len(groups))
	for key := range groups {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	result := make([]WorkItemGroup, 0, len(keys)+1)
	for _, key := range keys {
		children := groups[key]
		sort.Slice(children, func(i, j int) bool { return children[i].ID < children[j].ID })
		result = append(result, WorkItemGroup{Parent: parents[key], Items: children})
	}
	if len(orphans) != 0 {
		sort.Slice(orphans, func(i, j int) bool { return orphans[i].ID < orphans[j].ID })
		result = append(result, WorkItemGroup{Items: orphans})
	}
	return result, nil
}

func containsText(values []string, value string) bool {
	for _, candidate := range values {
		if candidate == value {
			return true
		}
	}
	return false
}

func (p *Provider) ResolvePullRequestWorkItemIDs(ctx context.Context, options Options, repositories, pullRequestIDs []string, token Token) ([]string, error) {
	if len(repositories) == 0 {
		return nil, &Error{Kind: ErrorInvalidInput, Detail: "PR mode requires an explicit repository, or a project with configured AzureDevOpsRepository entries."}
	}
	ids := make([]string, 0)
	for _, pullRequestID := range pullRequestIDs {
		numericID, err := strconv.ParseInt(pullRequestID, 10, 64)
		if err != nil {
			return nil, &Error{Kind: ErrorInvalidInput, Detail: "Invalid pull request ID: " + pullRequestID, Cause: err}
		}
		type match struct {
			repository string
			ids        []string
		}
		matches := make([]match, 0)
		for _, repository := range repositories {
			workItemIDs, found, err := p.TryGetPullRequestWorkItemIDs(ctx, options, repository, numericID, token)
			if err != nil {
				return nil, err
			}
			if found {
				matches = append(matches, match{repository: repository, ids: workItemIDs})
			}
		}
		switch len(matches) {
		case 0:
			return nil, &Error{Kind: ErrorRequest, Detail: "Pull request #" + pullRequestID + " was not found in tested Azure DevOps repos: " + strings.Join(repositories, ", ")}
		case 1:
			ids = append(ids, matches[0].ids...)
		default:
			names := make([]string, len(matches))
			for index, value := range matches {
				names[index] = value.repository
			}
			return nil, &Error{Kind: ErrorInvalidInput, Detail: "Pull request #" + pullRequestID + " was found in multiple repos (" + strings.Join(names, ", ") + "). Specify the repository."}
		}
	}
	seen := make(map[string]struct{})
	unique := make([]string, 0, len(ids))
	for _, id := range ids {
		if _, exists := seen[id]; !exists {
			seen[id] = struct{}{}
			unique = append(unique, id)
		}
	}
	return unique, nil
}

func (p *Provider) ActiveChildItems(ctx context.Context, options Options, parentID string, token Token) ([]WorkItemSnapshot, error) {
	ids, err := p.GetRelatedWorkItemIDs(ctx, options, parentID, RelationHierarchyForward, token)
	if err != nil {
		return nil, err
	}
	seen := make(map[string]struct{})
	unique := make([]string, 0, len(ids))
	for _, id := range ids {
		if _, exists := seen[id]; !exists {
			seen[id] = struct{}{}
			unique = append(unique, id)
		}
	}
	items, err := p.GetWorkItemsBatch(ctx, options, unique, token)
	if err != nil {
		return nil, err
	}
	active := items[:0]
	for _, item := range items {
		if !IsFinalState(valueOrEmpty(item.Type), valueOrEmpty(item.State)) {
			active = append(active, item)
		}
	}
	return active, nil
}
